i6_models/parts/factored_hybrid/triphone.py

```diff
@@ -0,0 +1,69 @@
+__all__ = [
+    "FactoredTriphoneBlockV1Config",
+    "FactoredTriphoneBlockV1",
+]
+
+from dataclasses import dataclass
+from typing import Tuple
+
+import torch
+from torch import nn, Tensor
+
+from .diphone import FactoredDiphoneBlockV1, FactoredDiphoneBlockV2Config
+from .util import get_mlp
+
+
+@dataclass
+class FactoredTriphoneBlockV1Config(FactoredDiphoneBlockV2Config):
+    """
+    Attributes:
+        Same as the FactoredDiphoneBlockV2Config.
+    """
+
+
+class FactoredTriphoneBlockV1(FactoredDiphoneBlockV1):
+    """
+    Triphone FH model output block.
+
+    Consumes the output h(x) of a main encoder model and computes factored logits/probabilities
+    for p(c|l,h(x)), p(l|h(x)) and p(r|c,l,h(x)).
+    """
+
+    def __init__(self, cfg: FactoredTriphoneBlockV1Config):
+        super().__init__(cfg)
+
+        self.center_state_embedding = nn.Embedding(self.num_center, cfg.center_state_embedding_dim)
+        self.right_context_encoder = get_mlp(
+            num_input=cfg.num_inputs + cfg.center_state_embedding_dim + cfg.left_context_embedding_dim,
+            num_output=self.num_contexts,
+            hidden_dim=cfg.context_mix_mlp_dim,
+            num_layers=cfg.context_mix_mlp_num_layers,
+            dropout=cfg.dropout,
+            activation=cfg.activation,
+        )
+
+    def forward(
```

Comment:
Here we overload the `forward` method. Which means that still the `FactoredDiphoneBlockV1.forward_factored` method exists in this class. Should we maybe overload `forward_factored` instead and then the `forward` inherited from `FactoredDiphoneBlockV1` would automatically do the correct thing.. except for typing...  so.. maybe inherit from V2 instead?

Related question: `forward_joint` is also still active, but only forwards the full diphone output ignoring the right context. Should we disable that? should we warn users? should we implement full triphone just in case anyone was so bold to ever try it?

Reply:
Good point. I don't think we should inherit from V2, because then we have some duplicate work done when calling `super().forward_factored(...)`, but I should definitely override `forward_factored` instead, and then also update the typings for `forward()`, as in Diphone V2.

Reply:
Wrt. `forward_joint`, a 10ms triphone LS960 model has about 450k outputs. I don't think this is ever going to be feasible, in whatever combination, w/o some clever optimizations. I think we should leave it at the diphone level for now, as this also allows decoding the triphone-trained models. It's a bit unfortunate that the `forward_joint` name does not include the term "diphone" anywhere, so that this would be clearer.